chain-impl-mockchain/src/leadership/genesis/mod.rs

```diff
@@ -295,6 +295,64 @@ mod tests {
         );
     }
 
+    #[test]
+    pub fn test_leader_with_invalid_pool_id() {
+        // Arrange
+        let leader_election_parameters = LeaderElectionParameters::new();
+
+        let cb = ConfigBuilder::new()
+            .with_slots_per_epoch(leader_election_parameters.slots_per_epoch)
+            .with_active_slots_coeff(leader_election_parameters.active_slots_coeff_as_milli());
+
+        let mut ledger = LedgerBuilder::from_config(cb)
+            .build()
+            .expect("cannot build test ledger")
+            .ledger;
+
+        let mut pools: HashMap<_, _>  = std::iter::from_fn(|| {
+            let (pool_id, pool_vrf_private_key) = make_pool(&mut ledger);
+            Some((
+                pool_id,
+                (pool_vrf_private_key, 0, leader_election_parameters.value),
+            ))
+        }).take(leader_election_parameters.pools_count).collect();
+
+        let selection = make_leadership_with_pools(&ledger, &pools);
+
+        let (invalid_pool_id, invalid_pool_vrf_private_key) = make_pool(&mut ledger);
+
+        // Act
+        let mut date = ledger.date();
+        let mut flag = true;
+
+        for _i in 00..leader_election_parameters.slots_per_epoch{
+            println!("Slot #{:?}, date: {:?}", _i, date);
+            for (pool_id, (pool_vrf_private_key, _, _)) in pools.iter_mut(){
+                
+                let leader = selection.leader(pool_id, pool_vrf_private_key, date);
+                match leader.unwrap()
+                {
+                    None => {}
+                    Some(_) => {
+                        println!("Leader Slot #{:?}, date: {:?}", _i, date);
+                        
+                        let invalid_leader = selection.leader(&invalid_pool_id, &invalid_pool_vrf_private_key, date);
```

Comment:
Why is this run only on slots for which other some other pool is selected?

Reply:
The two events are not related. Whether some pool is selected for a slot does not influence other pools. An invalid pool should get `None` all of the times, the others *might* be selected with a certain probability deriving from their stakes and some chain parameters.

Reply:
I see your point now, thanks for the clarification! I removed the check on the valid pools and checked only that the invalid one was never returning the unexpected value within one epoch.
Let me know if you see further improvements :)